app/assets/javascripts/components/bulkUploadImport.jsx

```diff
@@ -0,0 +1,399 @@
+      invalid: false,
+      errorMessage: '',
+      imported: false,
+      success: false,
+      successMessage: '',
+      hostName: this.hostName,
+      hostId: this.hostId,
+      project: 'Select a Project',
+      projectId: null,
+      selectedBulkPath: '',
+      samples: [],
+      selectedSampleIndices: [],
+      createdSampleIds: []
+    }
+  }
+
+  componentDidMount() {
+    this.initializeSelectTag();
+    $(ReactDOM.findDOMNode(this.refs.projectSelect)).on('change',this.handleProjectChange);
+    $(ReactDOM.findDOMNode(this.refs.hostSelect)).on('change',this.handleHostChange);
+  }
+
+  selectSample(e) {
+    // current array of options
+    const sample_list = this.state.selectedSampleIndices
+    console.log(sample_list)
+
+    let index
+
+    // check if the check box is checked or unchecked
+    if (e.target.checked) {
+      // add the numerical value of the checkbox to options array
+      sample_list.push(+e.target.id)
+    } else {
+      // or remove the value from the unchecked checkbox from the array
+      index = sample_list.indexOf(+e.target.id)
+      sample_list.splice(index, 1)
+    }
+
+    // update the state with the new array of options
+    this.setState({ selectedSampleIndices: sample_list })
+  }
+
+  handleUploadSubmit(e) {
+    e.preventDefault();
+    this.bulkUploadSubmit();
+  }
+
+  handleImportSubmit(e) {
+    e.preventDefault();
+    this.clearError();
+    if(!this.isImportFormInvalid()) {
+      this.bulkUploadImport()
+    }
+  }
+
+  initializeSelectTag() {
+    $('select').material_select();
+  }
+
+  clearError() {
+    this.setState({
+      invalid: false,
+      success: false
+     })
+  }
+
+  gotoPage(path) {
+    location.href = `${path}`
+  }
+
+
+  handleProjectSubmit(e) {
+    e.preventDefault();
+    this.clearError();
+    if(!this.isProjectInvalid()) {
+      this.addProject()
+    }
+  }
+
+  addProject() {
+    var that = this;
+    axios.post('/projects.json', {
+      project: {
+        name: this.refs.new_project.value
+      },
+      authenticity_token: this.csrf
+    })
+    .then((response) => {
+      var newProjectList = that.state.allProjects.slice();
+      newProjectList.push(response.data);
+      that.setState({
+        allProjects: newProjectList,
+        project: response.data.name,
+        projectId: response.data.id,
+        success: true,
+        successMessage: 'Project added successfully'
+      }, () => {
+        this.refs.new_project.value = '';
+        that.initializeSelectTag();
+      });
+    })
+    .catch((error) => {
+      that.setState({
+        invalid: true,
+        errorMessage: 'Project exists already or is invalid'
+      })
+    });
+  }
+
+  isProjectInvalid() {
+    if (this.refs.new_project.value === '' && this.state.project === 'Select a project') {
+      this.setState({
+        invalid: true,
+        errorMessage: 'Please enter valid project name'
+      })
+      return true;
+    } else {
+      return false;
```

Comment:
it just seems a bit unexpected for a predicate function to modify state, especially only in one of the cases